In [1]:
import os
import pandas as pd
from collections import OrderedDict
import numpy as np

`os.path.join(path, *paths)`<br>
Join one or more path components intelligently. The return value is the concatenation of path and any members of *paths with exactly one directory separator (os.sep) following each non-empty part except the last, meaning that the result will only end in a separator if the last part is empty. If a component is an absolute path, all previous components are thrown away and joining continues from the absolute path component.

### Country Code Concordances for Merging Records

In [2]:
codes = pd.read_csv('code-concordance.csv')

In [3]:
codes.head()

WBcode       WBname COWabb  COWcode      COWname        Income  MENA
0    AFG  Afghanistan    AFG    700.0  Afghanistan           Low     0
1    ALB      Albania    ALB    339.0      Albania  Upper-Middle     0
2    DZA      Algeria    ALG    615.0      Algeria  Upper-Middle     1
3    ADO      Andorra    AND    232.0      Andorra          High     0
4    AGO       Angola    ANG    540.0       Angola  Upper-Middle     0

### World Bank World Development Indicators (WDI)
https://data.worldbank.org/data-catalog/world-development-indicators

In [4]:
wdi = pd.read_csv(os.path.join('WDI', '7da42276-af1e-4ecf-ba32-04108473c2e7_Data.csv'))

In [5]:
wdi.head()

Country Name Country Code    Time Time Code  \
0  Afghanistan          AFG  1960.0    YR1960   
1  Afghanistan          AFG  1961.0    YR1961   
2  Afghanistan          AFG  1962.0    YR1962   
3  Afghanistan          AFG  1963.0    YR1963   
4  Afghanistan          AFG  1964.0    YR1964   

  Net official flows from UN agencies, UNAIDS (current US$) [DT.NFL.UNAI.CD]  \
0                                                 ..                           
1                                                 ..                           
2                                                 ..                           
3                                                 ..                           
4                                                 ..                           

  Net official aid received (current US$) [DT.ODA.OATL.CD]  \
0                                                 ..         
1                                                 ..         
2                                                 ..         
3                                                 ..         
4                                                 ..         

  Net bilateral aid flows from DAC donors, United States (current US$) [DC.DAC.USAL.CD]  \
0                                           15000000                                      
1                                           30000000                                      
2                                           13000000                                      
3                                           32000000                                      
4                                           37000000                                      

  Net bilateral aid flows from DAC donors, Total (current US$) [DC.DAC.TOTL.CD]  \
0                                           15670000                              
1                                           33550000                              
2                                           15260000                              
3                                           34880000                              
4                                           44220000                              

  GDP (current US$) [NY.GDP.MKTP.CD]  \
0                        537777811.1   
1                        548888895.6   
2                        546666677.8   
3                        751111191.1   
4                        800000044.4   

  GDP growth (annual %) [NY.GDP.MKTP.KD.ZG]  \
0                                        ..   
1                                        ..   
2                                        ..   
3                                        ..   
4                                        ..   

                                          ...                                           \
0                                         ...                                            
1                                         ...                                            
2                                         ...                                            
3                                         ...                                            
4                                         ...                                            

  Educational attainment, at least completed lower secondary, population 25+, male (%) (cumulative) [SE.SEC.CUAT.LO.MA.ZS]  \
0                                                 ..                                                                         
1                                                 ..                                                                         
2                                                 ..                                                                         
3                                                 ..                                                                         
4                                                 ..                                                            

In [6]:
wdi_mortality = pd.read_csv(os.path.join('WDI', '14f3b5cc-5125-4012-9b9c-517473664bf8_Data.csv'))

In [7]:
wdi_mortality.columns

Index(['Country Name', 'Country Code', 'Time', 'Time Code',
       'Maternal mortality ratio (modeled estimate, per 100,000 live births) [SH.STA.MMRT]',
       'Mortality rate, infant (per 1,000 live births) [SP.DYN.IMRT.IN]',
       'Life expectancy at birth, female (years) [SP.DYN.LE00.FE.IN]',
       'Life expectancy at birth, male (years) [SP.DYN.LE00.MA.IN]',
       'Life expectancy at birth, total (years) [SP.DYN.LE00.IN]'],
      dtype='object')

In [8]:
wdi_mortality.head()

Country Name Country Code    Time Time Code  \
0  Afghanistan          AFG  1960.0    YR1960   
1  Afghanistan          AFG  1961.0    YR1961   
2  Afghanistan          AFG  1962.0    YR1962   
3  Afghanistan          AFG  1963.0    YR1963   
4  Afghanistan          AFG  1964.0    YR1964   

  Maternal mortality ratio (modeled estimate, per 100,000 live births) [SH.STA.MMRT]  \
0                                                 ..                                   
1                                                 ..                                   
2                                                 ..                                   
3                                                 ..                                   
4                                                 ..                                   

  Mortality rate, infant (per 1,000 live births) [SP.DYN.IMRT.IN]  \
0                                              245.7                
1                                              241.2                
2                                              236.9                
3                                              232.7                
4                                              228.5                

  Life expectancy at birth, female (years) [SP.DYN.LE00.FE.IN]  \
0                                             33.133             
1                                             33.586             
2                                              34.03             
3                                             34.469             
4                                             34.904             

  Life expectancy at birth, male (years) [SP.DYN.LE00.MA.IN]  \
0                                              31.58           
1                                             32.026           
2                                             32.466           
3                                             32.903           
4                                             33.338           

  Life expectancy at birth, total (years) [SP.DYN.LE00.IN]  
0                                        32.33756098        
1                                        32.78697561        
2                                        33.22892683        
3                                        33.66690244        
4                                        34.10190244

In [9]:
del wdi['Country Name']
del wdi['Time Code']
wdi = wdi.merge(wdi_mortality, on=['Country Code', 'Time'], how='left')

In [10]:
#map abbreviated col names to long col descripts
wdi_col_map = OrderedDict()
for col in wdi.columns:
    if '[' in col:
        col_parts = col.split('[')
        wdi_col_map[col_parts[1].strip(']')] = col_parts[0].strip()
    else:
        wdi_col_map['.'.join(col.split())] = col

In [11]:
wdi.columns = wdi_col_map.keys()
wdi.replace('..', np.nan, inplace=True)

In [12]:
wdi.head()

Country.Code    Time DT.NFL.UNAI.CD DT.ODA.OATL.CD DC.DAC.USAL.CD  \
0          AFG  1960.0            NaN            NaN       15000000   
1          AFG  1961.0            NaN            NaN       30000000   
2          AFG  1962.0            NaN            NaN       13000000   
3          AFG  1963.0            NaN            NaN       32000000   
4          AFG  1964.0            NaN            NaN       37000000   

  DC.DAC.TOTL.CD NY.GDP.MKTP.CD NY.GDP.MKTP.KD.ZG NY.GDP.PCAP.CD  \
0       15670000    537777811.1               NaN    59.77732651   
1       33550000    548888895.6               NaN    59.87815281   
2       15260000    546666677.8               NaN    58.49287383   
3       34880000    751111191.1               NaN    78.78275804   
4       44220000    800000044.4               NaN    82.20844386   

  SP.POP.DPND.YG      ...       SL.TLF.CACT.MA.ZS SP.POP.TOTL SL.UEM.TOTL.ZS  \
0    76.52247462      ...                     NaN     8996351            NaN   
1    77.52917821      ...                     NaN     9166764            NaN   
2    78.08540637      ...                     NaN     9345868            NaN   
3    78.37535115      ...                     NaN     9533954            NaN   
4    78.62537426      ...                     NaN     9731361            NaN   

  Country.Name Time.Code SH.STA.MMRT SP.DYN.IMRT.IN SP.DYN.LE00.FE.IN  \
0  Afghanistan    YR1960         NaN          245.7            33.133   
1  Afghanistan    YR1961         NaN          241.2            33.586   
2  Afghanistan    YR1962         NaN          236.9             34.03   
3  Afghanistan    YR1963         NaN          232.7            34.469   
4  Afghanistan    YR1964         NaN          228.5            34.904   

  SP.DYN.LE00.MA.IN SP.DYN.LE00.IN  
0             31.58    32.33756098  
1            32.026    32.78697561  
2            32.466    33.22892683  
3            32.903    33.66690244  
4            33.338    34.10190244  

[5 rows x 34 columns]

### Merge WDI data with Codes data to start Main DataFrame

In [13]:
df = wdi.merge(codes, left_on='Country.Code', right_on='WBcode', how='left')
# remove redundant columns
del df['Country.Code']
del df['Country.Name']

In [14]:
df.head()

Time DT.NFL.UNAI.CD DT.ODA.OATL.CD DC.DAC.USAL.CD DC.DAC.TOTL.CD  \
0  1960.0            NaN            NaN       15000000       15670000   
1  1961.0            NaN            NaN       30000000       33550000   
2  1962.0            NaN            NaN       13000000       15260000   
3  1963.0            NaN            NaN       32000000       34880000   
4  1964.0            NaN            NaN       37000000       44220000   

  NY.GDP.MKTP.CD NY.GDP.MKTP.KD.ZG NY.GDP.PCAP.CD SP.POP.DPND.YG  \
0    537777811.1               NaN    59.77732651    76.52247462   
1    548888895.6               NaN    59.87815281    77.52917821   
2    546666677.8               NaN    58.49287383    78.08540637   
3    751111191.1               NaN    78.78275804    78.37535115   
4    800000044.4               NaN    82.20844386    78.62537426   

  SH.CON.1524.FE.ZS ...  SP.DYN.LE00.FE.IN SP.DYN.LE00.MA.IN SP.DYN.LE00.IN  \
0               NaN ...             33.133             31.58    32.33756098   
1               NaN ...             33.586            32.026    32.78697561   
2               NaN ...              34.03            32.466    33.22892683   
3               NaN ...             34.469            32.903    33.66690244   
4               NaN ...             34.904            33.338    34.10190244   

  WBcode       WBname COWabb COWcode      COWname Income MENA  
0    AFG  Afghanistan    AFG   700.0  Afghanistan    Low  0.0  
1    AFG  Afghanistan    AFG   700.0  Afghanistan    Low  0.0  
2    AFG  Afghanistan    AFG   700.0  Afghanistan    Low  0.0  
3    AFG  Afghanistan    AFG   700.0  Afghanistan    Low  0.0  
4    AFG  Afghanistan    AFG   700.0  Afghanistan    Low  0.0  

[5 rows x 39 columns]

### Polity IV Data on Regime Type (Autocracy - Democracy) and Durability
http://www.systemicpeace.org/inscrdata.html

In [17]:
polity = pd.read_excel('p4v2016.xls')
polity = polity[['ccode', 'year', 'polity2', 'durable']]
# polity2 = time series version of polity score -10, most autocratic, to 10, most democratic
# durable = number of years since a country has had a regime change (change of 3 or more in polity2)

In [18]:
polity.head()

ccode  year  polity2  durable
0    700  1800     -6.0      NaN
1    700  1801     -6.0      NaN
2    700  1802     -6.0      NaN
3    700  1803     -6.0      NaN
4    700  1804     -6.0      NaN

### Merge with Main DataFrame

In [19]:
df = df.merge(polity, left_on=['COWcode', 'Time'], right_on=['ccode', 'year'], how='left')
del df['ccode']
del df['Time']

In [20]:
df.head()

DT.NFL.UNAI.CD DT.ODA.OATL.CD DC.DAC.USAL.CD DC.DAC.TOTL.CD NY.GDP.MKTP.CD  \
0            NaN            NaN       15000000       15670000    537777811.1   
1            NaN            NaN       30000000       33550000    548888895.6   
2            NaN            NaN       13000000       15260000    546666677.8   
3            NaN            NaN       32000000       34880000    751111191.1   
4            NaN            NaN       37000000       44220000    800000044.4   

  NY.GDP.MKTP.KD.ZG NY.GDP.PCAP.CD SP.POP.DPND.YG SH.CON.1524.FE.ZS  \
0               NaN    59.77732651    76.52247462               NaN   
1               NaN    59.87815281    77.52917821               NaN   
2               NaN    58.49287383    78.08540637               NaN   
3               NaN    78.78275804    78.37535115               NaN   
4               NaN    82.20844386    78.62537426               NaN   

  SH.CON.1524.MA.ZS   ...   WBcode       WBname COWabb COWcode      COWname  \
0               NaN   ...      AFG  Afghanistan    AFG     700  Afghanistan   
1               NaN   ...      AFG  Afghanistan    AFG     700  Afghanistan   
2               NaN   ...      AFG  Afghanistan    AFG     700  Afghanistan   
3               NaN   ...      AFG  Afghanistan    AFG     700  Afghanistan   
4               NaN   ...      AFG  Afghanistan    AFG     700  Afghanistan   

  Income MENA    year polity2 durable  
0    Low  0.0  1960.0   -10.0     NaN  
1    Low  0.0  1961.0   -10.0     NaN  
2    Low  0.0  1962.0   -10.0     NaN  
3    Low  0.0  1963.0   -10.0     NaN  
4    Low  0.0  1964.0    -7.0     0.0  

[5 rows x 41 columns]

### USAID Donor/Recipient Data
http://aiddata.org/data/aiddata-core-research-release-level-1-3-1

In [22]:
aid = pd.read_csv(os.path.join('AidData', 'AidDataCoreDonorRecipientYearPurpose_ResearchRelease_Level1_v3.1.csv'))

##### Subset to aid delivered either by the UN or the US

In [23]:
aid.donor.value_counts()

United States                                                                             56631
Germany                                                                                   51193
France                                                                                    50875
Spain                                                                                     38655
Canada                                                                                    35829
Norway                                                                                    31935
Japan                                                                                     31220
United Kingdom                                                                            29545
United Nations Children`s Fund (UNICEF)                                                   29254
Belgium                                                                                   28236
Netherlands                             

In [24]:
aid = aid[(aid['donor'] == 'United States') | (aid['donor'].str.contains('United Nations'))]
aid.donor.value_counts()

United States                                                                             56631
United Nations Children`s Fund (UNICEF)                                                   29254
United Nations Development Programme (UNDP)                                               14802
United Nations Population Fund (UNFPA)                                                     5836
Joint United Nations Programme on HIV/AIDS (UNAIDS)                                        2102
United Nations Economic Commission for Europe (UNECE)                                       232
United Nations Democracy Fund (UNDEF)                                                       220
United Nations High Commissioner for Refugees (UNHCR)                                       158
United Nations Peacebuilding Fund (UNPBF)                                                   151
United Nations Relief and Works Agency for Palestine Refugees in the Near East (UNRWA)       48
United Nations Economic and Social Commi

In [25]:
# aggregate
aid['donor_agg'] = 'United Nations'
aid.loc[aid['donor']=='United States', 'donor_agg'] = 'United States'
aid.donor_agg.value_counts()

United States     56631
United Nations    52809
Name: donor_agg, dtype: int64

##### Focus on aid impacting women, reproductive rights, and family planning

In [26]:
for k, v in aid.coalesced_purpose_name.value_counts().items():
    print(k, '-', v)

Strengthening civil society - 4405
Multisector aid - 4120
Social/ welfare services - 3564
Reproductive health care - 3416
STD control including HIV/AIDS - 3008
Economic and development policy/planning - 2995
Std control including hiv/aids - 2969
Legal and judicial development - 2633
Government administration - 2478
Basic health care - 2325
Material relief assistance and services - 2304
Sectors not specified - 2258
Primary education - 2257
Population policy and administrative management - 2148
Family planning - 1706
Food security programmes/food aid - 1582
Multisector aid for social services - 1577
Basic nutrition - 1549
Environmental policy and administrative management - 1543
Business support services and institutions - 1411
Disaster prevention and preparedness - 1371
Population policy and admin. mgmt - 1313
Agricultural development - 1295
Bio-diversity - 1206
Infectious disease control - 1142
Employment policy and administrative management - 1138
Administrative costs - 1091
Industria

####### Categories of note
- Reproductive health care - 3416
- STD control including HIV/AIDS - 3008
- Std control including hiv/aids - 2969
- Family planning - 1706
- Women in development - 771
- Women's equality organisations and institutions - 169
- Health, combination of general, basic, and population policy/reproductive health purposes - 136
- Population policies/ programmes and reproductive health, combinations of activities - 135
- FAMILY PLANNING 12
- WOMEN'S EQUALITY ORGANISATIONS AND INSTITUTIONS 1

In [27]:
aid['aid_focus'] = 'Other'
aid.loc[aid['coalesced_purpose_name'] == 'Reproductive health care', 'aid_focus'] = 'Reproductive health care'
aid.loc[(aid['coalesced_purpose_name'] == 'STD control including HIV/AIDS') |
        (aid['coalesced_purpose_name'] == 'Std control including hiv/aids'), 'aid_focus'] = 'STD/HIV/AIDS'
aid.loc[(aid['coalesced_purpose_name'] == 'Family planning') |
        (aid['coalesced_purpose_name'] == 'FAMILY PLANNING'), 'aid_focus'] = 'Family planning'
aid.loc[aid['coalesced_purpose_name'] == 'Women in development', 'aid_focus'] = 'Women in development'
aid.loc[(aid['coalesced_purpose_name'] == "Women's equality organisations and institutions") |
        (aid['coalesced_purpose_name'] == "WOMEN'S EQUALITY ORGANISATIONS AND INSTITUTIONS"), 'aid_focus'] = \
        'Womens equality organizations and institutions'
aid.loc[(aid['coalesced_purpose_name'] == 'Health, combination of general, basic, and population policy/reproductive health purposes') |
        (aid['coalesced_purpose_name'] == "Population policies/ programmes and reproductive health, combinations of activities"), 'aid_focus'] = \
        'General health including reproductive health'
aid.aid_focus.value_counts()

Other                                             97117
STD/HIV/AIDS                                       5977
Reproductive health care                           3416
Family planning                                    1718
Women in development                                771
General health including reproductive health        271
Womens equality organizations and institutions      170
Name: aid_focus, dtype: int64

In [28]:
aid.columns

Index(['donor', 'recipient', 'year', 'coalesced_purpose_code',
       'coalesced_purpose_name', 'commitment_amount_usd_constant_sum',
       'donor_agg', 'aid_focus'],
      dtype='object')

##### Make wide (country-year format)

In [29]:
aid_donor_agg = aid.pivot_table(index=['recipient', 'year'], columns='donor_agg', 
                       values='commitment_amount_usd_constant_sum', aggfunc=sum).reset_index()
aid_donor = aid.pivot_table(index=['recipient', 'year'], columns='donor', 
                       values='commitment_amount_usd_constant_sum', aggfunc=sum).reset_index()
del aid_donor['United States']
aid_focus = aid.pivot_table(index=['recipient', 'year'], columns='aid_focus', 
                       values='commitment_amount_usd_constant_sum', aggfunc=sum).reset_index()
aid_focus['Womens Aid'] = aid_focus[['STD/HIV/AIDS', 'Reproductive health care', 'Family planning', 
                                 'Women in development', 'General health including reproductive health', 
                                 'Womens equality organizations and institutions']].sum(axis=1)
aid_donor_focus = aid.pivot_table(index=['recipient', 'year'], columns=['donor_agg', 'aid_focus'], 
                       values='commitment_amount_usd_constant_sum', aggfunc=sum).reset_index()
aid_donor_focus.columns = ['{}{}'.format(a, '{}'.format('-'+b if b else '')) for a, b in aid_donor_focus.columns]
un_cols = [col for col in aid_donor_focus if 'United Nations' in col and 'Other' not in col]
aid_donor_focus['United Nations-Womens Aid'] = aid_donor_focus[un_cols].sum(axis=1)
us_cols = [col for col in aid_donor_focus if 'United States' in col and 'Other' not in col]
aid_donor_focus['United States-Womens Aid'] = aid_donor_focus[us_cols].sum(axis=1)
aid_wide = aid_donor_agg.merge(aid_donor, on=['recipient', 'year'], 
                               how='outer').merge(aid_focus, on=['recipient', 'year'], how='outer') \
                                .merge(aid_donor_focus, on=['recipient', 'year'], how='outer')

In [30]:
aid_wide.fillna(0, inplace=True)

In [31]:
aid_wide.head()

recipient  year  United Nations  United States  \
0  Afghanistan  1973             0.0     40767666.0   
1  Afghanistan  1974             0.0     35999152.0   
2  Afghanistan  1975             0.0     27871564.0   
3  Afghanistan  1976             0.0      8688297.0   
4  Afghanistan  1977             0.0     16124956.0   

   Joint United Nations Programme on HIV/AIDS (UNAIDS)  \
0                                                0.0     
1                                                0.0     
2                                                0.0     
3                                                0.0     
4                                                0.0     

   United Nations Children`s Fund (UNICEF)  \
0                                      0.0   
1                                      0.0   
2                                      0.0   
3                                      0.0   
4                                      0.0   

   United Nations Democracy Fund (UNDEF)  \
0                                    0.0   
1                                    0.0   
2                                    0.0   
3                                    0.0   
4                                    0.0   

   United Nations Development Programme (UNDP)  \
0                                          0.0   
1                                          0.0   
2                                          0.0   
3                                          0.0   
4                                          0.0   

   United Nations Economic Commission for Europe (UNECE)  \
0                                                0.0       
1                                                0.0       
2                                                0.0       
3                                                0.0       
4                                                0.0       

   United Nations Economic and Social Commission for Asia and the Pacific (UNESCAP)  \
0                                                0.0                                  
1                                                0.0                                  
2                                                0.0                                  
3                                                0.0                                  
4                                                0.0                                  

             ...             \
0            ...              
1            ...              
2            ...              
3            ...              
4            ...              

   United Nations-Womens equality organizations and institutions  \
0                                                0.0               
1                                                0.0               
2                                                0.0               
3                                                0.0               
4                                                0.0               

   United States-Family planning  \
0                            0.0   
1                            0.0   
2                            0.0   
3                            0.0   
4                            0.0   

   United States-General health including reproductive health  \
0                                                0.0            
1                                                0.0            
2                                                0.0            
3                                                0.0            
4                                                0.0            

   United States-Other  United States-Reproductive health care  \
0           40767666.0                                     0.0   
1           35999152.0                                     0.0   
2           27871564.0                                     0.0   
3            8688297.0                                     0.0   
4           16124956.0                                     0.0  

##### Map recipient to country names from main data

In [32]:
recip_map_list = []
for recip in aid.recipient.unique():
    recip_dict = {'AIDname': recip}
    if recip in codes.WBname.tolist():
        recip_dict['WBname'] = codes.loc[codes['WBname']==recip, 'WBname'].values[0]
    else:
        recip_dict['WBname'] = None
    if recip in codes.COWname.tolist():
        recip_dict['COWname'] = codes.loc[codes['COWname']==recip, 'COWname'].values[0]
    else:
        recip_dict['COWname'] = None
    recip_map_list.append(recip_dict)
recip_map = pd.DataFrame(recip_map_list)

In [33]:
recip_map = recip_map.loc[~((recip_map.AIDname.str.contains('Regional')) | 
                            (recip_map.AIDname.str.contains('Bilateral')) | 
                            (recip_map.AIDname.str.contains('Unspecified')) |
                            (recip_map.AIDname.str.contains('Multi')) |
                            (recip_map.AIDname == 'no value') |
                            (recip_map.AIDname == 'Global') |
                            (recip_map.AIDname == 'European Commission'))]

In [34]:
recip_map[recip_map[['COWname', 'WBname']].isnull().all(axis=1)]

AIDname COWname WBname
21                Bosnia-Herzegovina    None   None
36     Congo, Democratic Republic of    None   None
37                Congo, Republic of    None   None
39                     Cote d`Ivoire    None   None
73     Korea, Democratic Republic of    None   None
104                             Niue    None   None
117              Sao Tome & Principe    None   None
128                      South Sudan    None   None
130                St. Kitts & Nevis    None   None
132          St.Vincent & Grenadines    None   None
143                Trinidad & Tobago    None   None
152                         Viet Nam    None   None
160           Palestinian Adm. Areas    None   None
167            Serbia and Montenegro    None   None
169                         Anguilla    None   None
170                     Cook Islands    None   None
182  Micronesia, Federated States of    None   None
183                       Montserrat    None   None
189                       St. Helena    None   None
190                          Tokelau    None   None
191         Turks and Caicos Islands    None   None
201                   Cayman Islands    None   None
206                 Hong Kong, China    None   None
213                            Macao    None   None
215                Northern Marianas    None   None
219                     Soviet Union    None   None
224              Virgin Islands (UK)    None   None

In [35]:
manual_map_list = [
    {
    'AIDname': 'Congo, Democratic Republic of',
    'WBname': 'Congo, Dem. Rep.',
    'COWname': 'Democratic Republic of the Congo'
    }, 
    {
    'AIDname': 'Congo, Republic of',
    'WBname': 'Congo, Rep.',
    'COWname': 'Congo'
    }, 
    {
    'AIDname': 'Cote d`Ivoire',
    'WBname': "Cote d'Ivoire",
    'COWname': 'Ivory Coast'
    },
    {
    'AIDname': 'Sao Tome & Principe',
    'WBname': 'Sao Tome and Principe',
    'COWname': 'Sao Tome and Principe'
    }, 
    {
    'AIDname': 'Micronesia',
    'WBname': 'Micronesia, Fed. Sts.',
    'COWname': 'Federated States of Micronesia'
    }, 
    {
    'AIDname': 'Viet Nam',
    'WBname': 'Vietnam',
    'COWname': 'Vietnam'
    }, 
    {
    'AIDname': 'Bosnia-Herzegovina',
    'WBname': 'Bosnia and Herzegovina',
    'COWname': 'Bosnia and Herzegovina'
    }, 
    {
    'AIDname': 'Korea, Democratic Republic of',
    'WBname': 'Korea, Democratic Republic of',
    'COWname': 'North Korea'
    }, 
    {
    'AIDname': 'St. Kitts & Nevis',
    'WBname': 'St. Kitts and Nevis',
    'COWname': 'St. Kitts and Nevis'
    }, 
    {
    'AIDname': 'St.Vincent & Grenadines',
    'WBname': 'St. Vincent and the Grenadines',
    'COWname': 'St. Vincent and the Grenadines'
    }, 
    {
    'AIDname': 'Trinidad & Tobago',
    'WBname': 'Trinidad and Tobago',
    'COWname': 'Trinidad and Tobago'
    }
]
manual_map = pd.DataFrame(manual_map_list)
manual_map

AIDname                           COWname  \
0   Congo, Democratic Republic of  Democratic Republic of the Congo   
1              Congo, Republic of                             Congo   
2                   Cote d`Ivoire                       Ivory Coast   
3             Sao Tome & Principe             Sao Tome and Principe   
4                      Micronesia    Federated States of Micronesia   
5                        Viet Nam                           Vietnam   
6              Bosnia-Herzegovina            Bosnia and Herzegovina   
7   Korea, Democratic Republic of                       North Korea   
8               St. Kitts & Nevis               St. Kitts and Nevis   
9         St.Vincent & Grenadines    St. Vincent and the Grenadines   
10              Trinidad & Tobago               Trinidad and Tobago   

                            WBname  
0                 Congo, Dem. Rep.  
1                      Congo, Rep.  
2                    Cote d'Ivoire  
3            Sao Tome and Principe  
4            Micronesia, Fed. Sts.  
5                          Vietnam  
6           Bosnia and Herzegovina  
7    Korea, Democratic Republic of  
8              St. Kitts and Nevis  
9   St. Vincent and the Grenadines  
10             Trinidad and Tobago

In [36]:
for i, row in manual_map.iterrows():
    recip_map.loc[recip_map['AIDname']==row['AIDname'], 'COWname'] = row['COWname']
    recip_map.loc[recip_map['AIDname']==row['AIDname'], 'WBname'] = row['WBname']

In [37]:
aid_wide = aid_wide.merge(recip_map, left_on='recipient', right_on='AIDname', how='left')
aid_wide = aid_wide.loc[~aid_wide[['COWname', 'WBname']].isnull().all(axis=1)]
aid_wide.head()

recipient  year  United Nations  United States  \
0  Afghanistan  1973             0.0     40767666.0   
1  Afghanistan  1974             0.0     35999152.0   
2  Afghanistan  1975             0.0     27871564.0   
3  Afghanistan  1976             0.0      8688297.0   
4  Afghanistan  1977             0.0     16124956.0   

   Joint United Nations Programme on HIV/AIDS (UNAIDS)  \
0                                                0.0     
1                                                0.0     
2                                                0.0     
3                                                0.0     
4                                                0.0     

   United Nations Children`s Fund (UNICEF)  \
0                                      0.0   
1                                      0.0   
2                                      0.0   
3                                      0.0   
4                                      0.0   

   United Nations Democracy Fund (UNDEF)  \
0                                    0.0   
1                                    0.0   
2                                    0.0   
3                                    0.0   
4                                    0.0   

   United Nations Development Programme (UNDP)  \
0                                          0.0   
1                                          0.0   
2                                          0.0   
3                                          0.0   
4                                          0.0   

   United Nations Economic Commission for Europe (UNECE)  \
0                                                0.0       
1                                                0.0       
2                                                0.0       
3                                                0.0       
4                                                0.0       

   United Nations Economic and Social Commission for Asia and the Pacific (UNESCAP)  \
0                                                0.0                                  
1                                                0.0                                  
2                                                0.0                                  
3                                                0.0                                  
4                                                0.0                                  

      ...       United States-Other  United States-Reproductive health care  \
0     ...                40767666.0                                     0.0   
1     ...                35999152.0                                     0.0   
2     ...                27871564.0                                     0.0   
3     ...                 8688297.0                                     0.0   
4     ...                16124956.0                                     0.0   

   United States-STD/HIV/AIDS  United States-Women in development  \
0                         0.0                                 0.0   
1                         0.0                                 0.0   
2                         0.0                                 0.0   
3                         0.0                                 0.0   
4                         0.0                                 0.0   

   United States-Womens equality organizations and institutions  \
0                                                0.0              
1                                                0.0              
2                                                0.0              
3                                                0.0              
4                                                0.0              

   United Nations-Womens Aid  United States-Womens Aid      AIDname  \
0                        0.0                       0.0  Afghanistan   
1                        0.0                       0.0  Afghanistan   
2                        0.0                       0.0  Afghanistan   
3                    

In [38]:
del aid_wide['recipient']

### Merge with Main DataFrame

In [39]:
# first on COWname, then WBname
df1 = df.merge(aid_wide, on=['COWname', 'year'], how='left')
del df1['WBname_y']
df1.rename(columns={'WBname_x': 'WBname'}, inplace=True)
df2 = df.merge(aid_wide, on=['WBname', 'year'], how='left')
del df2['COWname_y']
df2.rename(columns={'COWname_x': 'COWname'}, inplace=True)

In [40]:
df1.columns

Index(['DT.NFL.UNAI.CD', 'DT.ODA.OATL.CD', 'DC.DAC.USAL.CD', 'DC.DAC.TOTL.CD',
       'NY.GDP.MKTP.CD', 'NY.GDP.MKTP.KD.ZG', 'NY.GDP.PCAP.CD',
       'SP.POP.DPND.YG', 'SH.CON.1524.FE.ZS', 'SH.CON.1524.MA.ZS',
       'per_si_allsi.cov_pop_tot', 'per_sa_allsa.cov_pop_tot',
       'per_lm_alllm.cov_pop_tot', 'SE.SEC.CUAT.LO.ZS', 'SE.SEC.CUAT.LO.FE.ZS',
       'SE.SEC.CUAT.LO.MA.ZS', 'SE.PRM.CUAT.FE.ZS', 'SE.PRM.CUAT.MA.ZS',
       'SE.PRM.CUAT.ZS', 'SH.HIV.INCD.ZS', 'IT.NET.USER.ZS',
       'SL.TLF.CACT.FE.ZS', 'SL.TLF.CACT.MA.ZS', 'SP.POP.TOTL',
       'SL.UEM.TOTL.ZS', 'Time.Code', 'SH.STA.MMRT', 'SP.DYN.IMRT.IN',
       'SP.DYN.LE00.FE.IN', 'SP.DYN.LE00.MA.IN', 'SP.DYN.LE00.IN', 'WBcode',
       'WBname', 'COWabb', 'COWcode', 'COWname', 'Income', 'MENA', 'year',
       'polity2', 'durable', 'United Nations', 'United States',
       'Joint United Nations Programme on HIV/AIDS (UNAIDS)',
       'United Nations Children`s Fund (UNICEF)',
       'United Nations Democracy Fund (UNDEF)',
  

In [41]:
df1.columns == df2.columns

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True], dtype=bool)

In [42]:
df3 = pd.concat([df1, df2])

In [43]:
df = df3.drop_duplicates()
print(len(df3), len(df))

26422 12927


In [44]:
df.head()

DT.NFL.UNAI.CD DT.ODA.OATL.CD DC.DAC.USAL.CD DC.DAC.TOTL.CD NY.GDP.MKTP.CD  \
0            NaN            NaN       15000000       15670000    537777811.1   
1            NaN            NaN       30000000       33550000    548888895.6   
2            NaN            NaN       13000000       15260000    546666677.8   
3            NaN            NaN       32000000       34880000    751111191.1   
4            NaN            NaN       37000000       44220000    800000044.4   

  NY.GDP.MKTP.KD.ZG NY.GDP.PCAP.CD SP.POP.DPND.YG SH.CON.1524.FE.ZS  \
0               NaN    59.77732651    76.52247462               NaN   
1               NaN    59.87815281    77.52917821               NaN   
2               NaN    58.49287383    78.08540637               NaN   
3               NaN    78.78275804    78.37535115               NaN   
4               NaN    82.20844386    78.62537426               NaN   

  SH.CON.1524.MA.ZS   ...   United States-Family planning  \
0               NaN   ...                             NaN   
1               NaN   ...                             NaN   
2               NaN   ...                             NaN   
3               NaN   ...                             NaN   
4               NaN   ...                             NaN   

  United States-General health including reproductive health  \
0                                                NaN           
1                                                NaN           
2                                                NaN           
3                                                NaN           
4                                                NaN           

  United States-Other United States-Reproductive health care  \
0                 NaN                                    NaN   
1                 NaN                                    NaN   
2                 NaN                                    NaN   
3                 NaN                                    NaN   
4                 NaN                                    NaN   

  United States-STD/HIV/AIDS United States-Women in development  \
0                        NaN                                NaN   
1                        NaN                                NaN   
2                        NaN                                NaN   
3                        NaN                                NaN   
4                        NaN                                NaN   

  United States-Womens equality organizations and institutions  \
0                                                NaN             
1                                                NaN             
2                                                NaN             
3                                                NaN             
4                                                NaN             

  United Nations-Womens Aid United States-Womens Aid AIDname  
0                       NaN                      NaN     NaN  
1                       NaN                      NaN     NaN  
2                       NaN                      NaN     NaN  
3                       NaN                      NaN     NaN  
4                       NaN                      NaN     NaN  

[5 rows x 79 columns]

In [45]:
df.columns

Index(['DT.NFL.UNAI.CD', 'DT.ODA.OATL.CD', 'DC.DAC.USAL.CD', 'DC.DAC.TOTL.CD',
       'NY.GDP.MKTP.CD', 'NY.GDP.MKTP.KD.ZG', 'NY.GDP.PCAP.CD',
       'SP.POP.DPND.YG', 'SH.CON.1524.FE.ZS', 'SH.CON.1524.MA.ZS',
       'per_si_allsi.cov_pop_tot', 'per_sa_allsa.cov_pop_tot',
       'per_lm_alllm.cov_pop_tot', 'SE.SEC.CUAT.LO.ZS', 'SE.SEC.CUAT.LO.FE.ZS',
       'SE.SEC.CUAT.LO.MA.ZS', 'SE.PRM.CUAT.FE.ZS', 'SE.PRM.CUAT.MA.ZS',
       'SE.PRM.CUAT.ZS', 'SH.HIV.INCD.ZS', 'IT.NET.USER.ZS',
       'SL.TLF.CACT.FE.ZS', 'SL.TLF.CACT.MA.ZS', 'SP.POP.TOTL',
       'SL.UEM.TOTL.ZS', 'Time.Code', 'SH.STA.MMRT', 'SP.DYN.IMRT.IN',
       'SP.DYN.LE00.FE.IN', 'SP.DYN.LE00.MA.IN', 'SP.DYN.LE00.IN', 'WBcode',
       'WBname', 'COWabb', 'COWcode', 'COWname', 'Income', 'MENA', 'year',
       'polity2', 'durable', 'United Nations', 'United States',
       'Joint United Nations Programme on HIV/AIDS (UNAIDS)',
       'United Nations Children`s Fund (UNICEF)',
       'United Nations Democracy Fund (UNDEF)',
  

In [46]:
df.to_csv('full_data.csv', index=False)

In [47]:
wdi_col_map = pd.DataFrame.from_dict(wdi_col_map, orient='index').reset_index()
wdi_col_map.columns = ['col_name', 'description']
wdi_col_map.to_csv('wdi_column_name_mapping.csv', index=False)